volatility3/framework/layers/crash.py

```diff
@@ -67,34 +72,96 @@ def __init__(self, context: interfaces.context.ContextInterface, config_path: st
-            segments.append((x.BasePage * 0x1000, offset * 0x1000, x.PageCount * 0x1000, x.PageCount * 0x1000))
-            # print("Segments {:x} {:x} {:x}".format(x.BasePage * 0x1000,
-            #                  offset * 0x1000,
-            #                  x.PageCount * 0x1000))
-            offset += x.PageCount
+        if self.dump_type == 0x1:
+            header = self.context.object(self._crash_table_name + constants.BANG + self.dump_header_name,
+                                         offset=0,
+                                         layer_name=self._base_layer)
+
+            offset = self.headerpages
+            header.PhysicalMemoryBlockBuffer.Run.count = header.PhysicalMemoryBlockBuffer.NumberOfRuns
+            for x in header.PhysicalMemoryBlockBuffer.Run:
+                segments.append((x.BasePage * 0x1000, offset * 0x1000, x.PageCount * 0x1000, x.PageCount * 0x1000))
+                offset += x.PageCount
+
+        elif self.dump_type == 0x05:
+            summary_header = self.get_summary_header()
+            first_bit = None  # First bit in a run
+            first_offset = 0  # File offset of first bit
+            last_bit_seen = 0  # Most recent bit processed
+            offset = summary_header.HeaderSize  # Size of file headers
+            buffer_char = summary_header.get_buffer_char()
+            buffer_long = summary_header.get_buffer_long()
+
+            for outer_index in range(0, ((summary_header.BitmapSize + 31) // 32)):
+                if buffer_long[outer_index] == 0:
+                    if first_bit is not None:
+                        last_bit = ((outer_index - 1) * 32) + 31
+                        segment_length = (last_bit - first_bit + 1) * 0x1000
+                        segments.append((first_bit * 0x1000, first_offset, segment_length, segment_length))
+                        first_bit = None
+                elif buffer_long[outer_index] == 0xFFFFFFFF:
+                    if first_bit is None:
+                        first_offset = offset
+                        first_bit = outer_index * 32
+                    offset = offset + (32 * 0x1000)
+                else:
+                    for inner_index in range(0, 32):
+                        bit_addr = outer_index * 32 + inner_index
+                        if (buffer_char[bit_addr >> 3] >> (bit_addr & 0x7)) & 1:
+                            if first_bit is None:
+                                first_offset = offset
+                                first_bit = bit_addr
+                            offset = offset + 0x1000
+                        else:
+                            if first_bit is not None:
+                                segment_length = ((bit_addr - 1) - first_bit + 1) * 0x1000
+                                segments.append((first_bit * 0x1000, first_offset, segment_length, segment_length))
+                                first_bit = None
+                last_bit_seen = (outer_index * 32) + 31
```

Comment:
Feels like we might want some byte alignment methods just so everything's in one please?  If they're useful we could pull them up into some shared utility code somewhere, just feels like a lot of +/- 32s and 31s going on whereas it would be nicer to developers to say "bit_aligned_to_word` or something?